R/utils-latency.R

```diff
@@ -0,0 +1,171 @@
+#' offset each relevant column by it's appropriate latency
+#' works for either adjusting aheads or lags
+#' @param shift_cols a tibble which must have the columns `column`, the name of
+#'   the column to adjust, `latency` the latency of the original column relative
+#'   to the `as_of` date, `new_name`, the names in `column` adjusted by the
+#'   latencies `latency`
+#' @param new_data just what is says
+#' @param keys the variables which are used as keys
+#' @keywords internal
+extend_either <- function(new_data, shift_cols, keys) {
+  shifted <-
+    shift_cols %>%
+    select(-shifts, -effective_shift) %>%
+    pmap(\(original_name, latency, new_name) {
+      epi_shift_single(
+        x = new_data,
+        col = original_name,
+        shift_val = latency,
+        newname = new_name,
+        key_cols = keys
+      )
+    }) %>%
+    map(\(x) zoo::na.trim(x)) %>% # TODO need to talk about this
```

Comment:
Two things about this:
1. Zoo dependency. Seems we don't currently have this. are we trying to minimize package dependencies? This is the only function I use from it
2. This shears `NA`'s off the end of the column. We have a step that explicitly drops `NA`'s, so I'm not sure how much we're trying to avoid removing `NA`'s otherwise. Not sure how to do this correctly otherwise though.

Reply:
- by default, `na.trim` removes from both the beginning and the end
- side note: `\(` is R >= 4.1.  We were recently talking about not trying to support versions 3.* but not sure about 4.0.
- this function from latency processing code lying around may be helpful:
```r
n_true_at_tail = function(x) {
  match(FALSE, rev(x), nomatch = length(x) + 1L) - 1L
}
# So then to remove NAs at tail you could do the following (or maybe an S3 version of this):
remove_na_tail <- function(x) {
  x[seq_len(length(x) - n_true_at_tail(is.na(x)))]
}

remove_na_tail(c(NA, NA, 1:10, NA, 1:10, NA, NA))
remove_na_tail(1:100) # ALTREP
remove_na_tail(1:100 + 1) # non-ALTREP
remove_na_tail(integer(0L))
remove_na_tail(rep(NA_integer_, 5))
```

Reply:
> [oops just repeating Daniel here] 

yeah, threading in github issues can be hard to follow, especially when there are a lot of comments >.<

could you add some docs to those functions?